Enable AMF Decoder, HWContext AMF, vpp_amf filter

Hotactioncop opened the pull request:
Adds  hwcontext_amf, enabling a shared AMF context for encoders, 
decoders, and AMF-based filters, without copy to the host memory.

Benefits:
 - Optimizations for direct video memory access from CPU
 - Significant performance boost in full AMF pipelines with filters
 - Integration of GPU filters like VPP, Super Resolution, and 
    Compression Artefact Removal(in future plans)
 - VCN power management control for decoders.
 - Ability to specify which VCN instance to use for decoding 
   (like for encoder)
 - AMD will soon introduce full AMF API for multimedia accelerator MA35D
   - With AMF API, integration will be much easier: 
      GPU and the accelerator will have the same API
   - including encoder, decoder, scaler, color converter

Tested on:

- [x] Windows 10+  (via MinGW)
- [x] Ubuntu Linux

galad87 commented:
I have a feeling it would be better to wait until the patches are integrated into FFmpeg, and after #6446 is merged.

stickz commented:
Thanks so much for this! I'm running the AMF decoder with SVT-AV1 on my AMD Ryzen 9 9950x. It works great! I just had to update my bios and disable PBO to fix the BSOD problem.

Hotactioncop commented:
Asking for patch review.
What changed since last PR:
Updated to latest HandBrake master.
Updated contriib/ffmpeg patches to latest HWcontext, which is implemented into FFmpeg master.
Implemented GUI checkbox to enable AMF decoder on Linux.
Fixed crashes on Linux for copy_plane when default bytewidth larger than the source frame linesize.

Hotactioncop committed:
Enable AMF Decoder, HWContext AMF, vpp_amf filter

galad87 commented on libhb/common.c:
```diff
@@ -4508,6 +4508,7 @@ static void job_setup(hb_job_t * job, hb_title_t * title)
     job->qsv.decode                = !!(title->video_decode_support &
                                         HB_DECODE_SUPPORT_QSV);
 #endif
+    memset(&job->amf, 0, sizeof(job->amf));
```
hb_job_t is already zeroed on creation, this shouldn't be needed.

Hotactioncop replied:
Thanks, this will be applied in the upcoming fix commit.

galad87 commented on libhb/decavcodec.c:
```diff
@@ -1508,6 +1525,16 @@ int reinit_video_filters(hb_work_private_t * pv)
             hb_dict_set(settings, "format", hb_value_string(av_get_pix_fmt_name(pv->job->input_pix_fmt)));
             hb_avfilter_append_dict(filters, "scale_cuda", settings);
         }
+        else if ((pv->frame->hw_frames_ctx && pv->job && pv->job->hw_pix_fmt == AV_PIX_FMT_AMF_SURFACE) &&
```
A rotate filter is needed in this function too, if not all those video recorded on phones with a rotation metadata will fail to encode properly

Hotactioncop replied:
Currently, AMF Hardware filters do not support rotation. 
I will add an extra condition if (pv->title->rotation == HB_ROTATION_0) in the upcoming fix commit.

galad87 commented on libhb/decavcodec.c:
```diff
@@ -1829,20 +1875,21 @@ static int decodeFrame( hb_work_private_t * pv, packet_info_t * packet_info )
                 // HWAccel falled back to the software decoder
                 av_frame_ref(pv->frame, pv->hw_frame);
                 av_frame_unref(pv->hw_frame);
-                if (ret < 0)
```
Why?

Hotactioncop replied:
Accidentally removed, will be returned in the next fix commit.

galad87 commented on libhb/decavcodec.c:
```diff
@@ -1829,20 +1875,21 @@ static int decodeFrame( hb_work_private_t * pv, packet_info_t * packet_info )
-    } while (ret >= 0);
+        if(send_packet)
+        {
+            av_usleep(100);// wait 0.1 ms before resubmitting
```
Please don't add random usleep around.

Hotactioncop replied:
A sleep is required if the submit input is full, to prevent unnecessary CPU usage.

galad87 commented on libhb/decavcodec.c:
```diff
@@ -1927,15 +1975,32 @@ static int decavcodecvInit( hb_work_object_t * w, hb_job_t * job )
+        if(name != NULL)
+        {
+            pv->codec = avcodec_find_decoder_by_name(name);
+            if(pv->job) // no need to copy output for preview
```
I feel this is not a good way, what if we start to use hw filters for previews in the future?

Hotactioncop replied:
A reasonable remark. Will be removed in the next fix commit.

galad87 commented on libhb/decavcodec.c:
```diff
@@ -2553,6 +2618,21 @@ static int decavcodecvInfo( hb_work_object_t *w, hb_work_info_t *info )
         }
     }
 #endif
+#if HB_PROJECT_FEATURE_VCE
```
The other hardware encoders check if the hwaccel falled back to software decoder, wouldn't that work for amfdec too?

galad87 commented on libhb/handbrake/common.h:
```diff
@@ -848,6 +848,15 @@ struct hb_job_s
+    // AMF-specific settings
+    struct
+    {
+        int num_sw_filters;
```
Do we need to store these numbers?

Hotactioncop replied:
We need to store this variable because it is used to check whether software filters are present. 
If this variable is not 0, it means that software filters are present, and the pipeline can't be fully hardware-accelerated. However, since HandBrake is designed for maximum compatibility, we use int instead of bool.

galad87 commented on libhb/vce_common.c:
```diff
@@ -221,5 +316,178 @@ int hb_vce_av1_available()
+{
+    return job->hw_pix_fmt == AV_PIX_FMT_AMF_SURFACE;
+}
+hb_buffer_t *  hb_vce_copy_avframe_to_video_buffer(hb_job_t *job, AVFrame *frame, AVRational time_base)
```
Is this function needed? Looks the same as hb_avframe_to_video_buffer

galad87 replied:
The hw frame to sw frame conversion happens before, so there shouldn't be any need to check for sw_filters or whatever.

galad87 commented on libhb/decavcodec.c:
```diff
@@ -1927,15 +1975,32 @@ static int decavcodecvInit( hb_work_object_t * w, hb_job_t * job )
+    if(w->hw_device_ctx)
     {
-        pv->codec = avcodec_find_decoder_by_name(pv->qsv.codec_name);
+        const char* name = hb_vce_decode_get_codec_name(w->codec_param);
```
Would't be better to use the hwaccel framework like nvdec and videotoolboxdec? Don't try to copy the QSV implementation, because it was added before hwaccel, and it's doing custom things that hopefully will go away soon.

Hotactioncop replied:
The HWContext and AMF decoder is not hardware accelerated, it is an offload decoder.
Implementation of HWaccel AMF in hw_configs for default codecs was blocked by ffmpeg maintainers:
https://patchwork.ffmpeg.org/project/ffmpeg/patch/20240214015515.1027-2-ovchinnikov.dmitrii@gmail.com/

galad87 replied:
The issue with standalone hw decoders is that they are missing features, and are never in sync with the built-in parsers and hwaccel decoders. For example HDR dynamic metadata are ignored, and others metadata too. It's a pain because you never know what works and what doesn't.

sr55 replied:
That ultimately leads to a bad user experience when features that should be transparent and just work, don't.

galad87 reviewed (commented):
Added a bunch of random comments, didn't test this yet. But it would be better to replicate what NVenc/NVdec are doing (even if it was an hit and run and no one is keeping it working), instead of QSV.

Geoknysis commented:
There may be a bug that requires fixing before the feature is fully functional. Per the below, the output is sub-optimal and the work speed is slower. Below each image is the respective activity log for the encode. I also noticed that with the `amfdecode` option selected, when selecting the `constant framerate` video option, HandBrake crashes. 

I used the `official h265 1080p preset` for each. I compiled HandBrake with https://github.com/Hotactioncop/HandBrake/tree/Enable_AMF_decode source. 
 
![official_h265_vcn_1080p_amfdecode](https://github.com/user-attachments/assets/f19233c9-fdf7-41ab-86a0-b90d1b3d222c)
[Activity.log (official_h265_vcn_1080p_amfdecode).txt](https://github.com/user-attachments/files/19032946/Activity.log.official_h265_vcn_1080p_amfdecode.txt)

![official_h265_vcn_1080p_no_amfdecode](https://github.com/user-attachments/assets/b3f4ef22-af75-4c0b-867a-5cca6aef3d7e)
[Activity.log (official_h265_vcn_1080p_no_amfdecode).txt](https://github.com/user-attachments/files/19032952/Activity.log.official_h265_vcn_1080p_no_amfdecode.txt)

Hotactioncop commented:
> There may be a bug that requires fixing before the feature is fully functional. Per the below, the output is sub-optimal and the work speed is slower. Below each image is the respective activity log for the encode. I also noticed that with the `amfdecode` option selected, when selecting the `constant framerate` video option, HandBrake crashes.
> 
> I used the `official h265 1080p preset` for each. I compiled HandBrake with https://github.com/Hotactioncop/HandBrake/tree/Enable_AMF_decode source.
> 

Thank you for testing this PR! 
I noticed that you are seeing different frames compared to my results. 
To help me investigate and reproduce the issue, could you please provide the following details?

Input video file (or a short sample that reproduces the issue).
Your software setup, including:
- Linnux version
- GPU model
- Driver version
- Any specific build configuration or flags used
- Your HandBrake settings, especially the encoding parameters you used.

With my setup and a prepared video file matching the characteristics in your log file, I haven't encountered any errors. 
Having your exact input and environment details will help me debug this properly.
Thanks!

Mine181 commented:
any update on this merge request?

Hotactioncop commented:
Thanks for following up! 
The patch didn’t fit anymore after recent HandBrake updates, so It is rewritten to match the new code. 
It is currently under testing, and I will update the PR shortly.

Hotactioncop commented:
Work on this feature set has continued in follow-up PR: [#7698](https://github.com/HandBrake/HandBrake/pull/7698). 
This one is superseded.
